misk/src/main/kotlin/misk/web/extractors/QueryStringParameterProcessor.kt

```diff
@@ -0,0 +1,85 @@
+package misk.web.extractors
+
+import java.math.BigDecimal
+import java.math.BigInteger
+import kotlin.reflect.KParameter
+import kotlin.reflect.KType
+import kotlin.reflect.full.createType
+
+/**
+ * A query string parameter can be two things: a *primative* or a List<*primative*>
+ *     This class figures out which of the two is being represented based on the KParameter
+ *     used, as well as finding the KType of the primative.
+ */
+internal class QueryStringParameterProcessor constructor(parameter: KParameter) {
+    private val name: String? = parameter.name
+    private val isList: Boolean = parameter.type.classifier?.equals(List::class) ?: false
+    private val baseType: KType = if (isList) parameter.type.arguments.get(0).type!! else parameter.type
+
+    fun extractFunctionArgumentValue(parameterStrings: List<String>): Any? {
+        if (parameterStrings.isEmpty()) {
+            return null
+        }
+
+        if (isList) {
+            return parameterStrings.map { extractParameterValue(it) }
+        }
+        return extractParameterValue(parameterStrings.first())
+    }
+
+    private fun extractParameterValue(strValue: String): Any? {
+        try {
+            return when (baseType) {
+                stringType -> strValue
+                stringTypeNullable -> strValue
+                intType -> strValue.toInt()
+                intTypeNullable -> strValue.toInt()
+                longType -> strValue.toLong()
+                longTypeNullable -> strValue.toLong()
+                bigIntegerType -> strValue.toBigInteger()
+                bigIntegerTypeNullable -> strValue.toBigInteger()
+                floatType -> strValue.toFloat()
+                floatTypeNullable -> strValue.toFloat()
+                doubleType -> strValue.toDouble()
+                doubleTypeNullable -> strValue.toDouble()
+                bigDecimalType -> strValue.toBigDecimal()
+                bigDecimalTypeNullable -> strValue.toBigDecimal()
+                byteType -> strValue.toByte()
+                byteTypeNullable -> strValue.toByte()
+                byteArrayType -> strValue.toByteArray()
+                byteArrayTypeNullable -> strValue.toByteArray()
+                booleanType -> strValue.toBoolean()
+                booleanTypeNullable -> strValue.toBoolean()
+                else -> null
+            }
+        } catch(e: NumberFormatException) {
+            throw IllegalArgumentException(
```

Comment:
Use string interpolation for messages e.g

`"invalid parameter format for $name : $strValue"`

Reply:
Yeah, I forgot Kotlin could do that until a little while ago